File: sdk/python/examples/61_github_coding_agent_chained.py
#!/usr/bin/env python3
# Copyright (c) 2025 Agentspan
# Licensed under the MIT License. See LICENSE file in the project root for details.

"""GitHub Coding Agent — issue to PR pipeline.

Deploys and serves a three-stage pipeline:
  1. Fetch open issue, create branch (CLI tools: gh, git)
  2. Code fix + QA review (SWARM: coder <-> qa_tester)
  3. Create pull request (CLI tool: gh)

Run:
    python github_coding_agent.py          # Deploy + serve
    agentspan run github_pipeline "..."    # Trigger (from another terminal)

Requirements:
    - Agentspan server running
    - GITHUB_TOKEN stored: agentspan credentials set --name GITHUB_TOKEN
    - gh CLI installed
"""

from agentspan.agents import Agent, AgentRuntime, Strategy
from agentspan.agents.cli_config import CliConfig
from agentspan.agents.gate import TextGate
from agentspan.agents.handoff import OnTextMention

REPO = "agentspan-ai/codingexamples"
MODEL = "anthropic/claude-sonnet-4-6"

# ── Stage 1: Fetch issues ─────────────────────────────────────────

git_fetch_issues = Agent(
    name="git_fetch_issues",
    model=MODEL,
    max_tokens=8192,
    instructions=f"""\
You fetch ONE open issue from {REPO} and push an empty branch.

Step 1 — run this command:
  gh issue list --repo {REPO} --state open --limit 5
If no issues, respond: NO_OPEN_ISSUES

Step 2 — pick an issue, then run this ONE compound command (shell=true):
  TMPDIR=$(mktemp -d) && gh repo clone {REPO} "$TMPDIR" && cd "$TMPDIR" && git checkout -b fix/issue-<N> && git push -u origin fix/issue-<N> && echo "DONE"

Step 3 — respond with ONLY these 4 lines (NO tool calls):
  REPO: {REPO}
  BRANCH: fix/issue-<N>
  ISSUE: #<N> <title>
  SUMMARY: <one-sentence description>

RULES:
- Do NOT create files, commits, or pull requests.
- After step 2, you MUST stop using tools entirely. Just output text.
- You have at most 5 turns total.
""",
    cli_config=CliConfig(
        allowed_commands=["gh", "git", "mktemp"],
        allow_shell=True,
        timeout=60,
    ),
    credentials=["GITHUB_TOKEN", "GH_TOKEN"],
    max_turns=8,
    gate=TextGate("NO_OPEN_ISSUES"),
)

# ── Stage 2: Coding + QA (SWARM) ──────────────────────────────────

coder = Agent(
    name="coder",
    model=MODEL,
    max_tokens=60000,
    credentials=["GITHUB_TOKEN", "GH_TOKEN"],
    instructions="""\
You are a senior developer. Clone the repo, check out the branch,
implement the fix, commit, push. Then say HANDOFF_TO_QA with REPO/BRANCH/CHANGES.
""",
    local_code_execution=True,
)

qa_tester = Agent(
    name="qa_tester",
    model=MODEL,
    credentials=["GITHUB_TOKEN", "GH_TOKEN"],
    instructions="""\
You are a QA engineer. Clone the repo, review changes, run tests.
If bugs found: say HANDOFF_TO_CODER with what to fix.
If good: say QA_APPROVED with REPO/BRANCH/SUMMARY.
""",
    local_code_execution=True,
    max_tokens=60000,
    max_turns=5,
)

coding_qa = Agent(
    name="coding_qa",
    model=MODEL,
    instructions=(
        "Delegate to coder, then qa_tester. Loop until QA approves. "
        "Output REPO/BRANCH/SUMMARY when done."
    ),
    agents=[coder, qa_tester],
    strategy=Strategy.SWARM,
    handoffs=[
        OnTextMention(text="HANDOFF_TO_QA", target="qa_tester"),
        OnTextMention(text="HANDOFF_TO_CODER", target="coder"),
    ],
    max_turns=200,
    max_tokens=60000,
    timeout_seconds=6000,
)

# ── Stage 3: Create PR ────────────────────────────────────────────

git_push_pr = Agent(
    name="git_push_pr",
    model=MODEL,
    max_tokens=8192,
    max_turns=5,
    credentials=["GITHUB_TOKEN", "GH_TOKEN"],
    instructions="""\
Create a pull request. Run this ONE command (extract REPO, BRANCH, ISSUE from context):
  gh pr create --repo <REPO> --base main --head <BRANCH> --title "Fix <ISSUE>" --body "Fixes <ISSUE>"

After the command succeeds, STOP calling tools and respond with ONLY the PR URL.
""",
    cli_commands=True,
    cli_allowed_commands=["gh"],
)

# ── Pipeline ──────────────────────────────────────────────────────

pipeline = git_fetch_issues >> coding_qa >> git_push_pr

if __name__ == "__main__":
    with AgentRuntime() as rt:
        # Deploy: push definition to server (idempotent — safe to call every startup).
        # Can also be done via CLI: agentspan deploy examples.production.github_coding_agent
        # rt.deploy(pipeline)

        # Option A: Serve workers (production — blocks forever, run from outside)
        # rt.serve(pipeline)

        # Direct run for local development:
        # rt.run() handles deploy + workers internally, no serve needed.
        result = rt.run(pipeline, "Pick an open issue and create a PR.", timeout=240000)
        result.print_result()

        # Or trigger a deployed agent by name from any process:
        # result = rt.run("github_pipeline", "Pick an open issue and create a PR.")
        # result.print_result()

File: sdk/typescript/examples/61-github-coding-agent-chained.ts
/**
 * 61 - GitHub Coding Agent (Chained) — conditional sequential pipeline.
 *
 * Demonstrates:
 *   - Sequential pipeline with gate (conditional execution)
 *   - SWARM orchestration nested inside a pipeline stage
 *   - cliCommands for stages that only run CLI tools
 *   - localCodeExecution for stages that write/run code
 *
 * Architecture:
 *   pipeline = gitFetchIssues >> codingQA >> gitPushPR
 *
 * Requirements:
 *   - Conductor server running
 *   - AGENTSPAN_SERVER_URL=http://localhost:6767/api as environment variable
 *   - gh CLI authenticated
 */

import { Agent, AgentRuntime, OnTextMention, TextGate } from '../src/index.js';

const REPO = 'agentspan/codingexamples';
const MODEL = 'anthropic/claude-sonnet-4-6';

// -- Stage 1: Fetch issues ---------------------------------------------------

/** Stop when the agent has produced the structured output with issue details. */
function fetchDone(messages: unknown[]): boolean {
  const last = String(messages[messages.length - 1] ?? '');
  return ['REPO:', 'BRANCH:', 'ISSUE:', 'AUTHOR:', 'DETAILS:'].every(tag => last.includes(tag));
}

export const gitFetchIssues = new Agent({
  name: 'git_fetch_issues',
  model: MODEL,
  instructions:
    `You are a GitHub issue fetcher. Your ONLY job is to pick an issue and ` +
    `prepare a branch. Do NOT write any code or attempt to fix the issue — ` +
    `the next pipeline stage handles implementation.\n\n` +
    `1. List the 5 most recent open issues: gh issue list --repo ${REPO} --state open --limit 5\n` +
    `2. If there are NO open issues, output exactly: NO_OPEN_ISSUES\n` +
    `3. Pick the most suitable issue and fetch FULL details (body, author, labels):\n` +
    `   gh issue view <NUMBER> --repo ${REPO} --json number,title,body,author,labels\n` +
    `   You MUST run this — gh issue list only returns titles, not the issue body.\n` +
    `   Read the JSON output carefully and extract the author login and the COMPLETE body text.\n` +
    `4. Create and push a branch:\n` +
    `   - Create a temp dir: mktemp -d /tmp/fetch-XXXXXXXX\n` +
    `   - Clone ${REPO} into that dir\n` +
    `   - Create branch fix/issue-<NUMBER>\n` +
    `   - Push the empty branch: git push -u origin fix/issue-<NUMBER>\n` +
    `   - Delete the temp dir\n` +
    `5. Output ONLY these lines (NO tool calls):\n` +
    `       REPO: ${REPO}\n` +
    `       BRANCH: fix/issue-<NUMBER>\n` +
    `       ISSUE: #<NUMBER> <title>\n` +
    `       AUTHOR: <who opened the issue>\n` +
    `       DETAILS: <full issue body — preserve all requirements, acceptance criteria, and context>\n` +
    `       SUMMARY: <one-sentence description of the issue>\n\n` +
    `Include the COMPLETE issue body in DETAILS — the next stage needs it to implement the fix.`,
  cliConfig: { enabled: true, allowedCommands: ['gh', 'git', 'mktemp', 'rm'] },
  maxTurns: 20,
  stopWhen: fetchDone,
  gate: new TextGate({ text: 'NO_OPEN_ISSUES' }),
});

// -- Stage 2: Coding + QA (SWARM) -------------------------------------------

export const coderStage = new Agent({
  name: 'coder',
  model: MODEL,
  maxTokens: 60000,
  instructions:
    'You are a senior developer. Your input contains issue details from the previous stage\n' +
    'including REPO, BRANCH, ISSUE, AUTHOR, DETAILS, and SUMMARY.\n\n' +
    '1. Read the DETAILS field carefully — it contains the full issue body with requirements.\n' +
    '2. Create a fresh temp dir: mktemp -d /tmp/coder-XXXXXXXX\n' +
    '3. Clone the repo and check out the branch\n' +
    '4. Implement the fix according to ALL requirements in DETAILS.\n' +
    '5. Commit your changes with a descriptive message\n' +
    '6. Push: git push origin <BRANCH>\n' +
    '7. Delete the temp dir\n' +
    '8. Say HANDOFF_TO_QA followed by REPO/BRANCH/CHANGES lines',
  codeExecutionConfig: { enabled: true },
});

export const qaStage = new Agent({
  name: 'qa_tester',
  model: MODEL,
  instructions:
    'You are a QA engineer. Your task description contains REPO, BRANCH, and CHANGES.\n\n' +
    '1. Create a fresh temp dir and clone the repo/branch\n' +
    '2. Review the changed files and run tests\n' +
    '3. Delete the temp dir\n' +
    '4. If bugs: say HANDOFF_TO_CODER with details\n' +
    '5. If good: say QA_APPROVED followed by REPO/BRANCH/SUMMARY lines',
  codeExecutionConfig: { enabled: true },
  maxTokens: 60000,
  maxTurns: 5,
});

export const codingQA = new Agent({
  name: 'coding_qa',
  model: MODEL,
  instructions:
    'Your task description contains REPO, BRANCH, ISSUE, and SUMMARY. ' +
    'Delegate to coder to implement the fix, passing REPO, BRANCH, and the task details. ' +
    'Once coder completes, delegate to qa_tester. ' +
    'If QA does not pass, send it back to coder to fix. ' +
    'When QA approves, output ONLY these lines:\n' +
    '  REPO: <repo>\n' +
    '  BRANCH: <branch>\n' +
    '  SUMMARY: <what was implemented and verified>',
  agents: [coderStage, qaStage],
  strategy: 'swarm',
  handoffs: [
    new OnTextMention({ text: 'HANDOFF_TO_QA', target: 'qa_tester' }),
    new OnTextMention({ text: 'HANDOFF_TO_CODER', target: 'coder' }),
  ],
  maxTurns: 200,
  maxTokens: 60000,
  timeoutSeconds: 6000,
});

// -- Stage 3: Create PR ------------------------------------------------------

/** Stop when the agent has output a PR URL. */
function prDone(messages: unknown[]): boolean {
  const last = String(messages[messages.length - 1] ?? '');
  return last.includes('github.com') && last.includes('/pull/');
}

export const gitPushPR = new Agent({
  name: 'git_push_pr',
  model: MODEL,
  instructions:
    'You are a GitHub PR creator. Your input contains REPO, BRANCH, and SUMMARY from the previous stage.\n' +
    'The branch is already pushed -- your only job is to open a pull request.\n\n' +
    '1. Create the PR (use shell=true so quotes are handled correctly):\n' +
    '   gh pr create --repo <REPO> --base main --head <BRANCH> --title "<title>" --body "<summary>"\n' +
    '2. After the command succeeds, STOP calling tools and output ONLY the PR URL.',
  cliConfig: { enabled: true, allowedCommands: ['gh', 'git'], allowShell: true, timeout: 60 },
  maxTokens: 60000,
  maxTurns: 10,
  stopWhen: prDone,
});

// -- Pipeline ----------------------------------------------------------------

const pipeline = gitFetchIssues.pipe(codingQA).pipe(gitPushPR);

// Run the pipeline with streaming

// Only run when executed directly (not when imported for discovery)
async function main() {
  const runtime = new AgentRuntime();
  try {
    const result = await runtime.run(
    pipeline,
    `Pick the most suitable open issue on ${REPO} and implement a fix.`,
    );
    result.printResult();

    // Production pattern:
    // 1. Deploy once during CI/CD:
    // await runtime.deploy(pipeline);
    // CLI alternative:
    // agentspan deploy --package sdk/typescript/examples --agents git_fetch_issues
    //
    // 2. In a separate long-lived worker process:
    // await runtime.serve(pipeline);

    // Streaming alternative:
    // console.log('Starting pipeline: gitFetchIssues >> codingQA >> gitPushPR\n');
    // const agentStream = await runtime.stream(
    // pipeline,
    // `Pick the most suitable open issue on ${REPO} and implement a fix.`,
    // );

    // console.log(`Execution: ${agentStream.executionId}\n`);

    // for await (const event of agentStream) {
    // switch (event.type) {
    // case 'thinking':
    // console.log(`  [thinking] ${String(event.content).slice(0, 120)}...`);
    // break;
    // case 'tool_call':
    // console.log(`  [tool_call] ${event.toolName}(${JSON.stringify(event.args).slice(0, 100)})`);
    // break;
    // case 'tool_result':
    // console.log(`  [tool_result] ${event.toolName} -> ${String(event.result).slice(0, 200)}`);
    // break;
    // case 'error':
    // console.log(`  [error] ${event.content}`);
    // break;
    // case 'done':
    // console.log(`\n[done] Pipeline complete.`);
    // console.log(`Output: ${JSON.stringify(event.output).slice(0, 500)}`);
    // break;
    // default:
    // console.log(`  [${event.type}] ${JSON.stringify(event).slice(0, 150)}`);
    // }
    // }

    // const result = await agentStream.getResult();
    // console.log(`\nStatus: ${result.status}`);
    // console.log(`Tool calls: ${result.toolCalls.length}`);
  } finally {
    await runtime.shutdown();
  }
}

if (process.argv[1]?.endsWith('61-github-coding-agent-chained.ts') || process.argv[1]?.endsWith('61-github-coding-agent-chained.js')) {
  main().catch(console.error);
}
